src/MemoryTable.js

```diff
@@ -7,28 +8,67 @@ import ContextureMobx from './utils/contexture-mobx'
 import { componentForType } from './utils/schema'
 import { ResultTable, TypeMap } from './exampleTypes'
 
-export let memoryService = (records, { schema, debug } = {}) =>
-  Contexture({
-    debug,
-    // Hack to effectively set a default schema: if our tree root doesn't have
-    // a `schema` property, it will get the schema at key `undefined`.
-    schemas: { [schema]: { memory: { records } } },
-    providers: { memory: { ...memory, types: types() } },
-  })
-
-let MemoryTable = ({ data, fields, debug, include, ...props }) => {
-  let service = memoryService(data, { schema: 'data', debug })
+export let useMemoryTree = ({
+  records,
+  schema = 'data',
+  fields,
+  debug,
+  resultNode = {
+    pageSize: 50,
+  },
+  criteriaNodes = [],
+  childrenNodes = [],
+} = {}) => {
+  let include = _.map('field', fields)
+  let [memoryStorage] = React.useState({ records: [] })
   let [tree] = React.useState(
-    ContextureMobx({ service })({
+    ContextureMobx({
+      disableAutoUpdate: true,
+      service: Contexture({
+        debug,
+        schemas: { [schema]: { memory: memoryStorage } },
+        providers: { memory: { ...memory, types: types() } },
+      }),
+    })({
       key: 'root',
       schema: 'data',
       children: [
-        { key: 'results', type: 'results', include },
-        { key: 'criteria', type: 'group', children: [] },
+        { key: 'results', type: 'results', include, ...resultNode },
+        { key: 'criteria', type: 'group', children: criteriaNodes },
+        ...childrenNodes,
       ],
     })
   )
-  tree.refresh(['root'])
+
+  if (records !== memoryStorage.records) {
+    let updateMemory = async records => {
+      memoryStorage.records = await records
+      tree.refresh(['root'])
+    }
+    updateMemory(records)
+  }
+
+  return tree
+}
+
+let MemoryTable = ({
+  data,
+  fields,
+  debug,
+  resultNode,
+  criteriaNodes,
+  childrenNodes,
+  ...props
+}) => {
+  let tree = useMemoryTree({
+    records: data,
+    fields,
+    debug,
+    resultNode,
+    criteriaNodes,
+    childrenNodes,
+  })
+
```

Comment:
I think we can make it nicer and more composable with something like this

```javascript
let createResultTableTree = (criteria, results) => ({
  key: 'root',
  schema: 'data',
  children: [
     { key: 'results', type: 'results', include, ...results },
     { key: 'criteria', type: 'group', children: criteria },
  ]
}) 

let MemoryTable = ({ data, fields, criteria, results, ...props }) => {
  let tree = createResultTableTree(criteria, results)
  let client = useMemoryClient({ tree, data, fields })
  return <ResultTable path={['root', 'results']} criteria={['root', 'criteria']} />
}
```

That way users of `useMemoryClient` can pass whatever they want. For example:

```javascript
let client = useMemoryClient({ tree: { type: 'results', pageSize: 100 }, data, fields })
```

Reply:
I don't there is a need for the `memoryClient` of this kind, we actually want the hook that produces the tree. Also, this approach could produce a new tree every time, that we are trying to avoid

Reply:
After talking, we decided that although it's nicer to make this a bit more composable, we don't need the flexibility for now.